cmd/kubeadm/app/phases/upgrade/postupgrade.go

```diff
@@ -103,6 +106,28 @@ func PerformPostUpgradeTasks(client clientset.Interface, cfg *kubeadmapi.InitCon
 		errs = append(errs, err)
 	}
 
+	if err := PerformAddonsUpgrade(client, cfg, out); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errorsutil.NewAggregate(errs)
+}
+
+// PerformAddonsUpgrade performs the upgrade of the coredns and kube-proxy addons
+func PerformAddonsUpgrade(client clientset.Interface, cfg *kubeadmapi.InitConfiguration, out io.Writer) error {
+	if features.Enabled(cfg.FeatureGates, features.UpgradeAddonsAfterControlPlane) {
+		allUpgraded, unupgradedNodes, err := allControlPlaneInstancesUpgraded(client, cfg.NodeRegistration.Name)
+		if err != nil {
+			return errors.Wrapf(err, "failed to determine whether all the control plane instances have been upgraded")
+		}
+		if !allUpgraded {
+			fmt.Fprintf(out, "[upgrade/addons] skip upgrade addons because control plane instances %v have not been upgraded\n", unupgradedNodes)
```

Comment:
I am considering adding a warning message to notify users when the UpgradeAddonsAfterControlPlane feature gate is disabled and there are control planes that have not been upgraded. This will help users with multiple clusters understand the need to enable this feature gate and consider enabling it.

Reply:
Done